[5.2] Fix Permissions for Manually Running Scheduled Tasks

ditsuke opened the pull request:
Pull Request for Issue #36677.

### Summary of Changes
- [x] Explicitly disables the "run task" button on insufficient authorization.
- [x] Adds tooltip to give context on disabling the button.
- [ ] Change default permission level for running tasks/grant access to creator.
- [ ] Prevent manual runs from being publicly accessible regardless of permission levels. Ref: [#36453#issuecomment-1001793962](https://github.com/joomla/joomla-cms/issues/36453#issuecomment-1001793962) / @PhilETaylor 

### Testing Instructions
- Create a new task with an administrator (not superuser) account.
- Still as administrator, try to run it from the Scheduled tasks manager.
- Try to run other tasks created by the super user.
- Login as super user.
- Check that you can run all tasks.
- Thrash some task.
- Check the trashed tasks.

### Actual result BEFORE applying this Pull Request
As administrator (not superuser), there is no indication that the user is not authorized to run the task. The button is usable.
In all cases there is an error message.
Only a super user can run tasks.

### Expected result AFTER applying this Pull Request
As administrator (not superuser):
- The button is enabled for tasks created by yourself and you can run these tasks.
- The button is disabled for tasks which were created by the super user (.e.g those from a new installation). Reason for being disabled is available through a tooltip on hover.

As super user buttons of all tasks are enabled, and you can run these tasks.

When a task is trashed, the button is disabled, and a tool tip shows the reason.

### Documentation Changes Required
N/A

ditsuke committed:
Disable "run task" button on lack of authorization

- Explicitly disables the "run task" button on
  insufficient authorization.
- Adds tooltip to give context on disabling.

ditsuke committed:
Merge branch '4.1-dev' of github.com:joomla/joomla-cms 'fix-disabled-run-task-btn-on-no-auth-36677'

brianteeman commented:
How can it make sense for a user to be able to create a new task and setup lazy loading etc but not be able to run the task directly

ditsuke commented:
> How can it make sense for a user to be able to create a new task and setup lazy loading etc but not be able to run the task directly

Definitely, now that I think of it the creator should be able to run them manually as a default. I'll work on enabling that behavior before opening this PR.

brianteeman commented:
Probably need a new permission for that "Edit own"

ditsuke committed:
Allow creators of tasks the ability to run them

Task creators should now be able to run them even if they do not have
the `core.testrun` permission. The current permission checks should be
reviewed.

ditsuke commented on administrator/components/com_scheduler/src/Scheduler/Scheduler.php:
```diff
@@ -358,4 +359,30 @@ public function fetchTaskRecords(array $filters, array $listConfig): array
+	 *
+	 * @since __DEPLOY_VERSION__
+	 */
+	public static function isAuthorizedToRun(object $taskRecord, User $user): bool
```
@brianteeman what do you think about this check?

ditsuke committed:
Get `created_by` in default TasksModel query

We need this field in the authorization check. Begs the question, should
we just get all fields?

HLeithner commented:
This pull request has automatically rebased to 4.2-dev.

HLeithner commented:
This pull request has been automatically rebased to 4.3-dev.

HLeithner commented:
This pull request has been automatically rebased to 4.4-dev.

HLeithner commented:
This pull request has been automatically rebased to 5.2-dev.

Hackwar commented:
Can you please fix the conflicts?

richard67 committed:
Merge branch '5.2-dev' into fix-disabled-run-task-btn-on-no-auth-36677-update

richard67 committed:
Update ScheduleRunner.php

richard67 committed:
Revert unwanted change from merge conflict resolution

richard67 committed:
Merge branch '5.2-dev' into fix-disabled-run-task-btn-on-no-auth-36677

richard67 committed:
Fix error from conflict resolution

richard67 committed:
Merge branch 'fix-disabled-run-task-btn-on-no-auth-36677' of https://github.com/ditsuke/joomla-cms into fix-disabled-run-task-btn-on-no-auth-36677

richard67 committed:
Fix PHPCS

richard67 committed:
Again PHPCS

richard67 committed:
One more PHPCS

richard67 committed:
And another PHPCS

richard67 commented:
I was able to resolve the merge conflicts, which were mainly caused by PRs #40216 and #42746 .

It needs to be tested if this PR here and #42746 still work.

richard67 committed:
Merge branch '5.2-dev' into fix-disabled-run-task-btn-on-no-auth-36677

richard67 commented:
I have tested this item :white_check_mark: successfully on eb813f69cad69ecfaa068d0ecc1ed4538f9e3082<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36719">issues.joomla.org/tracker/joomla-cms/36719</a>.</sub>

dautrich commented:
I have tested this item :red_circle: unsuccessfully on eb813f69cad69ecfaa068d0ecc1ed4538f9e3082<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36719">issues.joomla.org/tracker/joomla-cms/36719</a>.</sub>

dautrich commented:
I've tested on J5.2.3 with a user belonging to user group "Administrator". 

- Without the patch applied, I was able to click on an existing task, but the task showed an error. 
- With the patch applied, the task execution button of the existing task  were "greyed" out and showed a tooltip "Not authorized".
![grafik](https://github.com/user-attachments/assets/e5cc9545-8e7b-485c-a068-cfc0fb723439)
- But with the patch applied, I was able to create additional tasks and run them:
![grafik](https://github.com/user-attachments/assets/4d8815e3-5c46-4150-9b09-d0636139c07d)

richard67 commented:
@dautrich Of course you shall be able to run your own tasks as user in the administrator group. But the ones created by the super user are disabled. To me it seems it is workings as intended.

richard67 commented:
P.S. Without the PR, one error is that you can not run tasks which you have created when you are an administrator. This is also fixed by this PR, that’s why you can create tasks and run them as administrator.

dautrich commented:
Then the testing instructions seem to be misleading. It says: "Create a new task with an admin (not superuser) account." And that's what I did.

richard67 commented:
Hmm I see the testing instructions are not precise.

richard67 commented:
@dautrich I've updated the testing instructions. Thanks for the hint.

richard67 commented:
P.S. I've updated again testing instructions and added a test for trashed tasks.

exlemor commented:
I have tested this item :white_check_mark: successfully on eb813f69cad69ecfaa068d0ecc1ed4538f9e3082<br /><br />I have tested this successfully.

(thanks for the updated testing instructions Richard!)


( Cool PR btw, thanks to all that made it possible ).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36719">issues.joomla.org/tracker/joomla-cms/36719</a>.</sub>

dautrich commented:
I have tested this item :white_check_mark: successfully on eb813f69cad69ecfaa068d0ecc1ed4538f9e3082<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36719">issues.joomla.org/tracker/joomla-cms/36719</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36719">issues.joomla.org/tracker/joomla-cms/36719</a>.</sub>

richard67 commented:
Thanks for testing.

richard67 committed:
Merge branch '5.2-dev' into fix-disabled-run-task-btn-on-no-auth-36677

Hackwar commented:
Thank you!
